### ISSUE TITLE
unhandledRejection in `flush` method

### ISSUE THREAD

**Issue by changgengli** [fluent/fluent-logger-forward-node#30]
When testing in in our production environment I noticed unhandledRejection with follow stack trace logged.

It seems this is around [here](https://github.com/fluent/fluent-logger-forward-node/blob/master/src/client.ts#L635
)

I also noticed EPIPE errors but the stack trace is not very helpful to pinpoint where exactly is wrong.

``` 
Error: write EPIPE
    at afterWriteDispatched (internal/stream_base_commons.js:156:25)
    at writeGeneric (internal/stream_base_commons.js:147:3)
    at Socket._writeGeneric (net.js:787:11)
    at Socket._write (net.js:799:8)
    at doWrite (_stream_writable.js:403:12)
    at writeOrBuffer (_stream_writable.js:387:5)
    at Socket.Writable.write (_stream_writable.js:318:11)
    at /srv/test/node_modules/@fluent-org/logger/build/src/socket.js:590:45
    at new Promise (<anonymous>)
    at FluentSocket.innerWrite (/srv/test/node_modules/@fluent-org/logger/build/src/socket.js:586:16)
    at FluentSocket.write (/srv/test/node_modules/@fluent-org/logger/build/src/socket.js:630:21)
    at FluentClient.sendNext (/srv/test/node_modules/@fluent-org/logger/build/src/client.js:459:42)
    at FluentClient.syncFlush (/srv/test/node_modules/@fluent-org/logger/build/src/client.js:321:36)
    at /srv/test/node_modules/@fluent-org/logger/build/src/client.js:336:30
    at processTicksAndRejections (internal/process/task_queues.js:79:11)
    at runNextTicks (internal/process/task_queues.js:66:3)
    at processImmediate (internal/timers.js:434:9)
    at process.topLevelDomainCallback (domain.js:138:15)
    at process.callbackTrampoline (internal/async_hooks.js:124:14)
```

Other stack trace:
```
Error: write EPIPE
    at WriteWrap.onWriteComplete [as oncomplete] (internal/stream_base_commons.js:94:16)
    at WriteWrap.callbackTrampoline (internal/async_hooks.js:126:14)

```

**Comment by jamiees2** [comment]
Hey, thanks for reporting! 

Could you by chance post the `Unhandled rejection` message? 

I think this might be due to `"error"` events always being emitted, I'd be interested if using `client.socketOn("error", () => {})` would prevent this. 

If it is, and I think the unhandled rejection message would confirm, I'll make a change to only emit the error event if there is a listener, to avoid this :)

**Comment by changgengli** [comment]
Not sure I follow.  My error handle looks like this. The above error is what logged by the unhandledRejection handler.

```
	process.on("unhandledRejection", (reason, promise) => {...})
```

**Comment by jamiees2** [comment]
Ah I see, I was just curious to see what the source of the error was, which Node gives you if you don't have the error handler set up. 

I think the problem is this line: https://github.com/fluent/fluent-logger-forward-node/blob/master/src/client.ts#L792, the promise rejects even though it is caught earlier, just another case of misunderstanding promises. I'll fix this tomorrow, but to confirm, you do not have acks enabled, correct? It looks like this bug would only manifest if acknowledgements are disabled.

**Comment by changgengli** [comment]
we don't have acks enable. Thanks for fixing the issue.

**Comment by jamiees2** [comment]
I merged my fix into v1.0.7, could you upgrade to that and check if the issue persists? :)

**Comment by jamiees2** [comment]
marking as resolved
